File: Segment Tree/3721. Longest Balanced Subarray II/solution.py
class SegmentTree:
    def __init__(self, size):
        self.size = size
        self.tree = [(0, 0)] * (4 * size)
        self.lazy = [0] * (4 * size)
        self.build(1, 0, size - 1)

    def build(self, node, start, end):
        if start == end:
            self.tree[node] = (0, 0)
            return
        mid = (start + end) // 2
        self.build(2 * node, start, mid)
        self.build(2 * node + 1, mid + 1, end)
        self.pull(node)

    def pull(self, node):
        self.tree[node] = (
            min(self.tree[2 * node][0], self.tree[2 * node + 1][0]),
            max(self.tree[2 * node][1], self.tree[2 * node + 1][1])
        )

    def push(self, node):
        if self.lazy[node] != 0:
            self.tree[2 * node] = (self.tree[2 * node][0] + self.lazy[node],
                                   self.tree[2 * node][1] + self.lazy[node])
            self.tree[2 * node + 1] = (self.tree[2 * node + 1][0] +
                                       self.lazy[node], self.tree[2 * node + 1][1] + self.lazy[node])
            self.lazy[2 * node] += self.lazy[node]
            self.lazy[2 * node + 1] += self.lazy[node]
            self.lazy[node] = 0

    def update(self, node, start, end, l, r, val):
        if start > end or start > r or end < l:
            return
        if l <= start and end <= r:
            self.tree[node] = (self.tree[node][0] + val,
                               self.tree[node][1] + val)
            self.lazy[node] += val
            return

        self.push(node)
        mid = (start + end) // 2
        self.update(2 * node, start, mid, l, r, val)
        self.update(2 * node + 1, mid + 1, end, l, r, val)
        self.pull(node)

    def update_range(self, l, r, val):
        self.update(1, 0, self.size - 1, l, r, val)

    def f_left(self, node, start, end, l, r, target):
        if start > end or start > r or end < l:
            return float('inf')
        if self.tree[node][0] > target or self.tree[node][1] < target:
            return float('inf')
        if start == end:
            return start if self.tree[node][0] == target else float('inf')

        self.push(node)
        mid = (start + end) // 2

        lft = self.f_left(2 * node, start, mid, l, r, target)
        if lft != float('inf'):
            return lft
        return self.f_left(2 * node + 1, mid + 1, end, l, r, target)

    def min_left(self, l, r, target):
        return self.f_left(1, 0, self.size - 1, l, r, target)

class Solution:
    def longestBalanced(self, nums: List[int]) -> int:
        n = len(nums)
        if n == 0: return 0

        last = {}
        pre = [-1] * n # the current occurrence of nums[i] is responsible for the range [pre[i]+1, i]
        for i in range(n):
            if nums[i] in last:
                pre[i] = last[nums[i]]
            last[nums[i]] = i

        st = SegmentTree(n)
        
        res = 0
        for j in range(n):
            chg = 1 if nums[j] % 2 == 0 else -1 # 1 or -1

            # Adds +1 (or -1) to all positions in range [pre[j]+1, j]
            # This marks: "if your subarray starts anywhere in [pre[j]+1, j] and ends at j, you now have one more distinct value of nums[j]"
            st.update_range(pre[j]+1, j, chg)

            i = st.min_left(0, j, 0)
            if i != float('inf'):
                res = max(res, j - i + 1)

        return res

File: Segment Tree/3721. Longest Balanced Subarray II/solution.cpp
// 1. Transform the problem: We want a subarray $[j, i]$ where $\text{DistinctEven}(j, i) - \text{DistinctOdd}(j, i) = 0$.
// 2. Maintain "Last Seen" Positions: As we iterate $i$ from $0$ to $n-1$:If nums[i] is even: it contributes $+1$ to any subarray starting after its previous occurrence.If nums[i] is odd: it contributes $-1$ to any subarray starting after its previous occurrence.
// 3. Segment Tree updates: At index $i$, we update the range $[previous[i]+1, i]$ with the contribution.The value at index $j$ in the segment tree represents the balance $\text{Even} - \text{Odd}$ for a subarray starting at $j$ and ending at $i$.
// 4. The Search: We need the smallest $j$ such that the value in the segment tree at index $j$ is $0$.
#include <vector>
#include <unordered_map>
#include <algorithm>

using namespace std;

class SegmentTree {
    int n;
    vector<int> minVal, maxVal, lazy;

public:
    SegmentTree(int n) : n(n), minVal(4 * n, 0), maxVal(4 * n, 0), lazy(4 * n, 0) {}

    void push(int v) {
        if (lazy[v] != 0) {
            minVal[2 * v] += lazy[v];
            maxVal[2 * v] += lazy[v];
            lazy[2 * v] += lazy[v];
            minVal[2 * v + 1] += lazy[v];
            maxVal[2 * v + 1] += lazy[v];
            lazy[2 * v + 1] += lazy[v];
            lazy[v] = 0;
        }
    }

    void update(int v, int tl, int tr, int l, int r, int add) {
        if (l > r) return;
        if (l == tl && r == tr) {
            minVal[v] += add;
            maxVal[v] += add;
            lazy[v] += add;
        } else {
            push(v);
            int tm = (tl + tr) / 2;
            update(2 * v, tl, tm, l, min(r, tm), add);
            update(2 * v + 1, tm + 1, tr, max(l, tm + 1), r, add);
            minVal[v] = min(minVal[2 * v], minVal[2 * v + 1]);
            maxVal[v] = max(maxVal[2 * v], maxVal[2 * v + 1]);
        }
    }

    int find_leftmost(int v, int tl, int tr, int target) {
        // If target 0 is not in the range [min, max], it's not here
        if (minVal[v] > target || maxVal[v] < target) return -1;
        if (tl == tr) return tl;
        
        push(v);
        int tm = (tl + tr) / 2;
        int res = find_leftmost(2 * v, tl, tm, target);
        if (res == -1) {
            res = find_leftmost(2 * v + 1, tm + 1, tr, target);
        }
        return res;
    }
};

class Solution {
public:
    int longestBalanced(vector<int>& nums) {
        int n = nums.size();
        unordered_map<int, int> lastIdx;
        SegmentTree st(n);
        int res = 0;

        for (int i = 0; i < n; i++) {
            int val = nums[i];
            int prev = lastIdx.count(val) ? lastIdx[val] : -1;
            int contrib = (val % 2 == 0) ? 1 : -1;

            // Add contribution only to subarrays where this instance is the FIRST occurrence from the left
            st.update(1, 0, n - 1, prev + 1, i, contrib);

            int j = st.find_leftmost(1, 0, n - 1, 0);
            if (j != -1 && j <= i) {
                res = max(res, i - j + 1);
            }
            lastIdx[val] = i;
        }
        return res;
    }
};